app/javascript/src/apps/mydb/elements/details/reactions/variationsTab/ReactionVariationsComponents.js

```diff
@@ -0,0 +1,513 @@
+import {
+  getReferenceMaterial, getCatalystMaterial, getFeedstockMaterial, getMolFromGram, getGramFromMol,
+  computeEquivalent, computePercentYield
+} from 'src/apps/mydb/elements/details/reactions/variationsTab/ReactionVariationsMaterials';
+import { parseNumericString } from 'src/utilities/MathUtils';
+import {
+  calculateGasMoles, calculateTON, calculateFeedstockMoles, calculateFeedstockVolume
+} from 'src/utilities/UnitsConversion';
+
+function RowToolsCellRenderer({
+  data: variationsRow, context
+}) {
+  const { reactionShortLabel, copyRow, removeRow } = context;
+  return (
+    <div>
+      <Badge bg="secondary" className="me-2">{getVariationsRowName(reactionShortLabel, variationsRow.id)}</Badge>
+      <ButtonGroup>
+        <Button size="xsm" variant="success" onClick={() => copyRow(variationsRow)}>
+          <i className="fa fa-clone" />
+        </Button>
+        <Button size="xsm" variant="danger" onClick={() => removeRow(variationsRow)}>
+          <i className="fa fa-trash-o" />
+        </Button>
+      </ButtonGroup>
+    </div>
+  );
+}
+
+RowToolsCellRenderer.propTypes = {
+  data: PropTypes.shape({
+    id: PropTypes.number.isRequired,
+  }).isRequired,
+  context: PropTypes.shape({
+    reactionShortLabel: PropTypes.string.isRequired,
+    copyRow: PropTypes.func.isRequired,
+    removeRow: PropTypes.func.isRequired,
+  }).isRequired,
+};
+
+function EquivalentParser({ data: variationsRow, oldValue: cellData, newValue }) {
+  let equivalent = parseNumericString(newValue);
+  if (equivalent < 0) {
+    equivalent = 0;
+  }
+  // Adapt mass to updated equivalent.
+  const referenceMaterial = getReferenceMaterial(variationsRow);
+  const referenceMol = getMolFromGram(referenceMaterial.mass.value, referenceMaterial);
+  const mass = getGramFromMol(referenceMol * equivalent, cellData);
+
+  // Adapt amount to updated equivalent.
+  const amount = getMolFromGram(mass, cellData);
+
+  return {
+    ...cellData,
+    mass: { ...cellData.mass, value: mass },
+    amount: { ...cellData.amount, value: amount },
+    equivalent: { ...cellData.equivalent, value: equivalent }
+  };
+}
+
+function PropertyFormatter({ value: cellData, colDef }) {
+  const { displayUnit } = colDef.entryDefs;
+  const valueInDisplayUnit = convertUnit(Number(cellData.value), cellData.unit, displayUnit);
+
+  return `${Number(valueInDisplayUnit).toPrecision(4)}`;
+}
+
+function PropertyParser({
+  oldValue: cellData, newValue, colDef
+}) {
+  const { currentEntry, displayUnit } = colDef.entryDefs;
+  let value = parseNumericString(newValue);
+  if (currentEntry !== 'temperature' && value < 0) {
+    value = 0;
+  }
+  value = convertUnit(value, displayUnit, cellData.unit);
+  const updatedCellData = { ...cellData, value };
+
+  return updatedCellData;
+}
+
+function MaterialFormatter({ value: cellData, colDef }) {
+  const { currentEntry, displayUnit } = colDef.entryDefs;
```

Comment:
can you check this error? , it happens when I click on duration button: cellData undefined

![image](https://github.com/user-attachments/assets/083c56b3-111a-4a4e-9982-5287b09f0048)